dbms/src/Storages/KVStore/MultiRaft/Disagg/CheckpointIngestInfo.cpp

```diff
@@ -168,42 +167,115 @@ void CheckpointIngestInfo::persistToLocal() const
     uni_ps->write(std::move(wb), DB::PS::V3::PageType::Local, nullptr);
     LOG_INFO(
         log,
-        "Successfully persist CheckpointIngestInfo, region_id={} peer_id={} remote_store_id={} region={}",
+        "Successfully persist CheckpointIngestInfo, region_id={} peer_id={} remote_store_id={} region={} segments={}",
         region_id,
         peer_id,
         remote_store_id,
-        region->getDebugString());
+        region->getDebugString(),
+        restored_segments.size());
 }
 
-void CheckpointIngestInfo::removeFromLocal(TMTContext & tmt, UInt64 region_id)
+static void removeFromLocal(TMTContext & tmt, UInt64 region_id)
 {
-    auto log = DB::Logger::get();
-    LOG_INFO(log, "Erase CheckpointIngestInfo from disk, region_id={}", region_id);
     auto uni_ps = tmt.getContext().getWriteNodePageStorage();
     UniversalWriteBatch del_batch;
     del_batch.delPage(
         UniversalPageIdFormat::toLocalKVPrefix(UniversalPageIdFormat::LocalKVKeyType::FAPIngestInfo, region_id));
     uni_ps->write(std::move(del_batch), PageType::Local);
 }
 
-// Like removeFromLocal, but is static and with check.
-bool CheckpointIngestInfo::forciblyClean(TMTContext & tmt, UInt64 region_id, bool pre_check)
+void CheckpointIngestInfo::deleteWrittenData(TMTContext & tmt, RegionPtr region, const DM::Segments & segments)
 {
-    if (!pre_check)
+    auto region_id = region->id();
+    auto & storages = tmt.getStorages();
+    auto keyspace_id = region->getKeyspaceID();
+    auto table_id = region->getMappedTableID();
+    auto storage = storages.get(keyspace_id, table_id);
+
+    auto log = DB::Logger::get("CheckpointIngestInfo");
+    if (storage && storage->engineType() == TiDB::StorageEngine::DT)
     {
-        CheckpointIngestInfo::removeFromLocal(tmt, region_id);
-        return true;
+        auto dm_storage = std::dynamic_pointer_cast<StorageDeltaMerge>(storage);
+        auto dm_context = dm_storage->getStore()->newDMContext(tmt.getContext(), tmt.getContext().getSettingsRef());
+        for (const auto & segment_to_drop : segments)
+        {
+            DM::WriteBatches wbs(*dm_context->storage_pool, dm_context->getWriteLimiter());
```

Comment:
Suggest collecting one WriteBatches for all segments and then use one write call to remove all persisted page

Reply:
What if a exception throws before we write the `wbs`? It will have all dmfiles deleted while themselves are still registered in pagestorage
```
void Segment::dropAsFAPTemp(const FileProviderPtr & file_provider, WriteBatches & wbs)
{
    delta->recordRemoveColumnFilesPages(wbs);
    stable->recordRemovePacksPages(wbs);
    wbs.writeAll();
    stable->drop(file_provider);
}
```

Reply:
Removing the stable files from disk but leaving the page_id to those dmfile on UniPS may make it fail to restore the StableVS (which tries to read the meta from dmfile) from removing next time. https://github.com/pingcap/tiflash/blob/8e90a26de37e039e37091af98681a3f3374f0a88/dbms/src/Storages/KVStore/MultiRaft/Disagg/CheckpointIngestInfo.cpp#L87-L91

You can do it this way:
* collect the stable files
* collect the page_ids in StableVS and remove them in one write batch
* call `DMFile::remove` (or more lightweight `DMFile::enableGC` and the file will be GCed later)

Reply:
Making that logic to another PR is LGTM, because it is kind of complicated and not effect the correctness of this PR.

Reply:
Consider it as a non-blocking issue, and use another pr to improve is LGTM.
I think we'd better first in all solve some compat issues for cse.